### ISSUE TITLE
Improve the "Generate Missing WebP Versions" Function in the Settings Page

### ISSUE THREAD

**Issue by joejoe04** [wp-media/imagify-plugin#549]
https://secure.helpscout.net/conversation/1434208379/241822

- Seems there may be a bug where it goes through the process of generating the missing WebP versions, and seems to be doing so successfully, but after the process is complete, the same number of missing WebP images is still reported.
- It's unclear whether the WebP versions are created but the message remains afterwards due to a bug or if the WebP versions are not actually created (even though it seems to be happening successfully during the process).
- It's also possible that backups for these images don't exist, but there's no indication that that is the case

**To Reproduce**
Unable to reproduce this from my end at the moment, but there are a fair amount of users who have this issue where clicking the "Generate Missing WebP Versions" button either doesn't work to create the WebP images, or it does work, but the message that the WebP versions are missing still remains afterward

**Expected behavior**
- Either the WebP versions should be generated and the message about missing WebP versions should be removed, or if the WebP images are unable to be created, there could be some indication of why and for which images this is the case.
- Also, when it lists the number of images with missing WebP versions, it would be very much helpful if it would also list which images it is referring to. 
- Would also be good if there was a way to know if backup images were missing for these images
  - When users have 10s of thousands of images and this isn't working, it's very hard to offer them a good solution because there's no way to know which images it's referring to without searching manually through all images for missing WebP versions and/or original backups.

**Comment by iCaspar** [comment]
Cannot reproduce (yet!).

However, in the case cited in the OP, the problem was specifically that the backed-up originals were missing. In such cases Imagify cannot generate WebP (or for that matter, re-generate any other sizes/compression levels) as the saved originals are required for any optimization.
Perhaps the UI could be more clear about that specific situation. It would require additional check for the existence of a backed-up original at the time of loading any given page of images in the media library, noting that "missing backup" and removing the links for any re-optimization. As it is, I think the behavior shown in the customer's screenshots is the fallback provisionally intended -- the popup says it cannot be done because the backups are missing.

**Comment by joejoe04** [comment]
Doing some additional testing, I think there actually could be at least a couple of issues happening, and I believe I may have found a way to reproduce at least some of them, including the issue where even though backup images do exist, the tool is still refusing to create the WebP versions and also the issue where the "you have N optimized images without WebP versions" is not updated after the process is completed.

To start, I upload 2 images to the Media Library with WebP creation disabled and allow them to optimize. After doing this, I see the settings page say there are 2 images with missing WebP versions. Then I enable WebP creation and delete the backup version of one of the images. 

When doing this and clicking "Generate Missing WebP Versions" I see this (even though it says there are 2 missing WebP versions, the tool displays that it is working on 0/1 images:

![image](https://user-images.githubusercontent.com/45705494/145212113-487a228b-a66c-47d1-9aed-a8f6ad6c68d1.png)

Even though this is technically correct, since there is no indication of why the image with the missing backup image is not being included, I think this is confusing for users. They see that there are 2 images with missing WebP versions, but only 1 is being worked on.

Not sure about a better option just yet, but this is something we could talk about if we also discuss options for how we might be able to indicate which specific images are missing their backups and/or WebP versions (if possible).

Once the process is finished, it reveals another issue, where the page does not update and it still says there are 2 images with missing WebP versions:

![image](https://user-images.githubusercontent.com/45705494/145212228-e6b54eb6-4fe9-42b7-8615-c871f4434c70.png)

Refreshing the page does update this, though:

![image](https://user-images.githubusercontent.com/45705494/145213504-04753fce-db94-4f65-b7d9-8f0344d055d0.png)

I imagine this could be solved relatively easily.

But that brings me to the more serious issue, where if I restore the formerly deleted backup for the remaining image without WebP to its proper backup location, I still get the popup saying, "Because the selected images did not have a backup copy, Imagify was unable to create WebP versions.":

![image](https://user-images.githubusercontent.com/45705494/145223287-337fc489-7b6e-4bd6-a3a2-a877e8cb31f3.png)

![image](https://user-images.githubusercontent.com/45705494/145223832-b298dbe8-c60e-4d9d-a79e-8bbbc807fd59.png)

![image](https://user-images.githubusercontent.com/45705494/145223467-5b3cd1d5-ed1b-4133-9528-b9b0dd5d695c.png)

![image](https://user-images.githubusercontent.com/45705494/145223569-b0274a66-b178-4709-8d33-42f0ea8c8fde.png)

With all this said, one more strange part about this is that I can only reproduce this sometimes. On my local site setup on my laptop, it happens most times I repeat this process (though there was one time where it worked as it should and generated the WebP versions).

However, when I tried reproducing the issue on the test site I have hosted online, things worked fine once I restored the backup image most of the time. But I kept repeating the process, and eventually I got the same thing to happen.

![image](https://user-images.githubusercontent.com/45705494/145237866-2dd4d065-ff18-402b-a086-7b610cdda43a.png)

![image](https://user-images.githubusercontent.com/45705494/145237997-b656c0e0-364e-4754-9491-ed6ed23c8733.png)

![image](https://user-images.githubusercontent.com/45705494/145238051-608d74f5-347b-46e8-baf7-7e75191cc3cf.png)

Here are the images I used just in case. I'm not sure how the specific images used might matter, but providing them just in case:
![creek](https://user-images.githubusercontent.com/45705494/145240718-4e5a6a8f-d849-4366-8381-0181205255c8.jpg)
![test](https://user-images.githubusercontent.com/45705494/145240832-4e0f0f3b-970f-4585-b95b-066c3c54d865.jpg)

Would anyone else like to try the process outlined above to see if they can reproduce the issue as well?

UPDATE: One more thing to add is that when I restore the backup image, in the Media Library, I do see the "Generate WebP versions" link become available, it does work to create the WebP versions, and the "1 optimized images without WebP versions" message in the settings page goes away:

![image](https://user-images.githubusercontent.com/45705494/145244588-8c25b42a-f84b-4196-b1cd-271689e25c2c.png)

**Comment by joejoe04** [comment]
Additionally, we have an issue where even though there are images listed as needing WebP versions generated, when the "Generate Missing WebP Versions" button is clicked it appears as though it's doing the processing to generate them, but after getting through all the images, there is still the same number of images with missing WebP versions reported.

I don't know for sure how to reproduce this issue, but I do know there is an example of it happening currently on http://mega.wp-rocket.me/divi/. If you go to the Imagify settings page, we see 2 images reported as missing WebP versions:

![image](https://user-images.githubusercontent.com/45705494/145241999-5ef5ddec-cfd1-4ef0-8270-a0f6e69c5d63.png)

If you then click the button, it appears to be processing 1 of them (seems that one of these must not have a backup version, so that would explain why that one is not having WebP versions generated):

![image](https://user-images.githubusercontent.com/45705494/145242258-55f52585-a99d-4cea-91dd-d262c980ceb7.png)

But after it's done, the same 2 images are listed (even after refreshing the page):

![image](https://user-images.githubusercontent.com/45705494/145242396-3a0c3ba0-dc3b-4a8d-a1dc-a7ee7de4abd4.png)

We believe at least one cause of this is when there are WELL DONE images that were optimized before the WebP feature was added to Imagify and that if you find which image this is referring to in the Media Library and clicked "Try Again", the WebP versions would likely be created in that case.

With that said, I'm not sure if there might also be other causes for this as well. I'm also not sure which image this is referring to, as this is occurring for an image that must have been uploaded and optimized some time ago.

But posting this in case we should investigate the reasons this could be happening (if there are any other causes than the one mentioned above) and to see if it might be possible to allow for the "Generate Missing WebP Versions" button work to create the WebP versions in these cases.

**Comment by iCaspar** [comment]
I'm going to put the `Epic` tag on this one, as it seems to be a combination of several issues around webp generation. We can chip away at it to identify the various sub-issues and try to make some progress.

**Issue 1:**
To start, I upload 2 images to the Media Library with WebP creation disabled and allow them to optimize. After doing this, I see the settings page say there are 2 images with missing WebP versions. Then I enable WebP creation and delete the backup version of one of the images.

When doing this and clicking "Generate Missing WebP Versions" I see this (even though it says there are 2 missing WebP versions, the tool displays that it is working on 0/1 images.

In this case, we know why the webp version is not generated, but "working on 0/1 images" messaging is misleading/confusing.

If we change this to say "working on 0/2 images" and then report afterward that 1 failed, that might be less confusing. Should be a [S] change.

Alternatively, we could scan the images missing webp and pre-count those with no backup, so that the first message would read like "You have 1 image without webp where Imagify can create a webp version." -- ignoring the ones we know we can't regenerate beforehand. May be even better. Would be more involved to implement - [S-M]

**Issue 2:**
Once the process is finished, it reveals another issue, where the page does not update and it still says there are 2 images with missing WebP versions.

Again, this should be a [S], as we just need to trigger a page refresh.

**Issue 3:**
If I restore the formerly deleted backup for the remaining image without WebP to its proper backup location, I still get the popup saying, "Because the selected images did not have a backup copy, Imagify was unable to create WebP versions."

We'll need to groom this separately. Especially as the problem is described as intermittent or perhaps system dependent.

**Issue 4:**
(Possibly related to 3)
The "Generate Missing WebP Versions" button is clicked it appears as though it's doing the processing to generate them, but after getting through all the images, there is still the same number of images with missing WebP versions reported.

Description above suggests that if for any reason a webp cannot be generated (backup missing, initial upload of pre-optimized preventing, etc, there is no record of which images failed or why. So we will need to:
(a) track and keep information from previous webp generation attempts; and
(b) use that info when generating the message about missing image counts (not counting them as possible items for generation). We can also find a way to be more clear about messaging what failed in the attempt. "Sorry Imagify could not generate webp version of the following images: ..." So at least the user knows what happened. 

This could potentially be a [L] issue. Further grooming will be needed to outline a solution here.

**Comment by iCaspar** [comment]
#652 Should cover issues 1 and 2 above. Turns out they are closely related.

**Comment by webprom** [comment]
Also looks that even when I'm trying to generate webp for a very old site without backup images, the process eat up all my credits, without generating anything.
